File: CSharp/572-SubtreeOfAnotherTree/572SubtreeOfAnotherTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 // Time Complexity: O(n * m) (Analyze Complexity Confirmed)
 // Space Complexity: O(n) (Analyze Complexity Confirmed)
public class Solution {
    public bool IsSubtree(TreeNode root, TreeNode subRoot) {
        if(subRoot is null) return true;
        if(root is null) return false;

        if(IsSameTrees(root, subRoot)) return true;

        return IsSubtree(root.left, subRoot) || IsSubtree(root.right, subRoot);
    }

    private bool IsSameTrees(TreeNode root, TreeNode subRoot) {
        if(root is null && subRoot is null) return true;

        if(root is not null &&
           subRoot is not null &&
           root.val == subRoot.val &&
           IsSameTrees(root.left, subRoot.left) &&
           IsSameTrees(root.right, subRoot.right)) return true;

        return false;
    }
}

File: Javascript/572-SubrootOfAnotherTree/572SubtreeOfAnotherTree.js
/*
Time Complexity: O(n * m)
where n is the number of nodes in the root tree
and m is the number of nodes in the subRoot tree

Space Commplexity: O(h + hs)
where h is the height of root
and hs is the height of subroot

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    const areEqual = (node1, node2) =>{
        if (!node1 && !node2) return true;
        if (!node1 || !node2 || node1.val !== node2.val) return false;
        return areEqual(node1.left, node2.left) && areEqual(node1.right, node2.right);
    }

    const dfs = node => {
        if (!node) return false;
        if (areEqual(node, subRoot)) return true;
        return dfs(node.left) || dfs(node.right);
    }
    return dfs(root);
};